src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,439 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import analysis.normed_space.basic
+
+/-!
+# M-structure
+
+A continuous projection P on a normed space X is said to be an L-projection if, for all `x` in `X`,
+$$
+∥x∥ = ∥P x∥ + ∥(1-P) x∥.
+$$
+The range of an L-projection is said to be an L-summand of X.
+
+A continuous projection P on a normed space X is said to be an M-projection if, for all `x` in `X`,
+$$
+∥x∥ = max(∥P x∥,∥(1-P) x∥).
+$$
+The range of an M-projection is said to be an M-summand of X.
+
+The L-projections and M-projections form Boolean algebras. When X is a Banach space, the Boolean
+algebra of L-projections is complete.
+
+Let `X` be a normed space with dual `X^*`. A closed subspace `M` of `X` is said to be an M-ideal if
+the topological annihilator `M^∘` is an L-summand of `X^*`.
+
+M-ideal, M-summands and L-summands were introduced by Alfsen and Effros in [alfseneffros1972] to
+study the structure of general Banach spaces. When `A` is a JB*-triple, the M-ideals of `A` are
+exactly the norm-closed ideals of `A`. When `A` is a JBW*-triple with predual `X`, the M-summands of
+`A` are exactly the weak*-closed ideals, and their pre-duals can be identified with the L-summands
+of `X`. In the special case when `A` is a C*-algebra, the M-ideals are exactly the norm-closed
+two-sided ideals of `A`, when `A` is also a W*-algebra the M-summands are exactly the weak*-closed
+two-sided ideals of `A`.
+
+## Implementation notes
+
+The approach to showing that the L-projections form a Boolean algebra is inspired by
+`measure_theory.measurable_space`.
+
+## References
+
+* [Behrends, M-structure and the Banach-Stone Theorem][behrends1979]
+* [Harmand, Werner, Werner, M-ideals in Banach spaces and Banach algebras][harmandwernerwerner1993]
+
+## Tags
+
+M-summand, M-projection, L-summand, L-projection, M-ideal, M-structure
+
+-/
+
+variables {M : Type*} [monoid M]
+
+/--
+A continuous linear map `P` on a normed space `X` is said to be a projection if it is idempotent.
+-/
+def is_projection (x : M) : Prop := x^2 = x
```

Comment:
Could this whole section on `is_projection` (i.e. everything before norms are mentioned) be moved, say to `src/algebra/ring/is_projection.lean`?

Reply:
Since this is for arbitrary monoids/rings, maybe `is_idempotent` could be a better name? Also, you might be already aware of that but just in case I'll mention that `linear_algebra/projection` has some results on projection maps, which maybe can be useful for you.

Reply:
@antoinelab01 Thanks - I'd wondered about that, but the name `is_idempotent` already seems to be used in a slightly different sense in lean's built in library [here](https://github.com/leanprover-community/lean/blob/master/library/init/algebra/classes.lean#L54).

Thanks for mentioning  `linear_algebra/projection` - I'd spotted that - it's likely to be needed for moving from projections to summands.

The other thing I've been wondering about is whether there's a single agreed way to define a partial ordering on projections for any algebra. I think probably not.

Reply:
I've broken out projections (idempotents) into a separate PR #13830.